x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java

```diff
@@ -0,0 +1,271 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.geometry.ShapeType;
+import org.elasticsearch.index.IndexVersions;
+import org.elasticsearch.index.mapper.GeoShapeQueryable;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryShardException;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.lucene.spatial.CartesianShapeDocValuesQuery;
+import org.elasticsearch.search.sort.NestedSortBuilder;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+import org.elasticsearch.xpack.ql.querydsl.query.Query;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypes.CARTESIAN_POINT;
+
+public class SpatialRelatesQuery extends Query {
+    private final String field;
+    private final ShapeRelation queryRelation;
+    private final Geometry shape;
+    private final DataType dataType;
+
+    public SpatialRelatesQuery(Source source, String field, ShapeRelation queryRelation, Geometry shape, DataType dataType) {
+        super(source);
+        this.field = field;
+        this.queryRelation = queryRelation;
+        this.shape = shape;
+        this.dataType = dataType;
+    }
+
+    @Override
+    public boolean containsNestedField(String path, String field) {
+        return false;
+    }
+
+    @Override
+    public Query addNestedField(String path, String field, String format, boolean hasDocValues) {
+        return null;
+    }
+
+    @Override
+    public void enrichNestedSort(NestedSortBuilder sort) {
+
+    }
+
+    @Override
+    public QueryBuilder asBuilder() {
+        return EsqlDataTypes.isSpatialGeo(dataType) ? new GeoShapeQueryBuilder() : new CartesianShapeQueryBuilder();
+    }
+
+    @Override
+    protected String innerToString() {
+        throw new IllegalArgumentException("SpatialRelatesQuery.innerToString() not implemented");
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field, queryRelation, shape, dataType);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+
+        SpatialRelatesQuery other = (SpatialRelatesQuery) obj;
+        return Objects.equals(field, other.field)
+            && Objects.equals(queryRelation, other.queryRelation)
+            && Objects.equals(shape, other.shape)
+            && Objects.equals(dataType, other.dataType);
+    }
+
+    public abstract class ShapeQueryBuilder implements QueryBuilder {
```

Comment:
Can we add a comment that is public for testing?

Reply:
Done